compiler-rt/cmake/Modules/CompilerRTUtils.cmake

```diff
@@ -388,6 +394,30 @@ macro(construct_compiler_rt_default_triple)
     set(COMPILER_RT_DEFAULT_TARGET_ARCH "i386")
   endif()
 
+  # If we are directly targeting a GPU we need to check that the compiler is
+  # compatible and pass some default arguments.
+  if(COMPILER_RT_DEFAULT_TARGET_ONLY)
+    if("${COMPILER_RT_DEFAULT_TARGET_ARCH}" MATCHES "amdgcn|nvptx")
+      # Ensure the compiler is a valid clang when building the GPU target.
+      set(req_ver "${LLVM_VERSION_MAJOR}.${LLVM_VERSION_MINOR}.${LLVM_VERSION_PATCH}")
+      if(LLVM_VERSION_MAJOR AND NOT (CMAKE_CXX_COMPILER_ID MATCHES "[Cc]lang" AND
+         ${CMAKE_CXX_COMPILER_VERSION} VERSION_EQUAL "${req_ver}"))
+        message(FATAL_ERROR "Cannot build compiler-rt for GPU. CMake compiler "
+                            "'${CMAKE_CXX_COMPILER_ID} ${CMAKE_CXX_COMPILER_VERSION}' "
+                            " is not 'Clang ${req_ver}'.")
+      endif()
+      set(COMPILER_RT_GPU_TARGET TRUE)
+    endif()
+
+    # Pass the necessary flags to make flag detection work.
+    if("${COMPILER_RT_DEFAULT_TARGET_ARCH}" MATCHES "amdgcn")
+      set(CMAKE_REQUIRED_FLAGS "${CMAKE_REQUIRED_FLAGS} -nogpulib")
+    elseif("${COMPILER_RT_DEFAULT_TARGET_ARCH}" MATCHES "nvptx")
+      # FIXME: This doesn't work with no CUDA installation.
+      set(CMAKE_REQUIRED_FLAGS "${CMAKE_REQUIRED_FLAGS} -march=sm_75")
+    endif()
+  endif()
```

Comment:
Can this be extracted somehow? I suspect that we need this in base-config-ix.cmake as well.

Reply:
What's the difference? I thought this was the one place that checked the default argument.

Reply:
I think for a regular build (not cross compiling directly) it will be fine since it will just use `clang` which should hopefully be supported by the user's system, this is mostly required for when the Runtimes build invoked it with `CMAKE_C_TARGET=nvptx64-nvidia-cuda`.

Reply:
Specifying `CMAKE_C_TARGET` doesn't imply cross-compilation though. I could use a x64 compiler, hosted on x64, and still specify the x64 target and that would not be cross-compilation.

Reply:
True, but I'm just going off of the expected case where the compiler has an implicit `--target=` version. I'm assuming the base logic just checks if it can target any of  them? The only logic I saw for that was in `base-config-ix.cmake` and it went off of the default target, so I'm unsure where else to modify.